scanner/decisions/0005-scannerv4-read-cvss-data.md

```diff
@@ -0,0 +1,40 @@
+# 0005 - Adapt Scanner V4 NVD CVSS data updating pipeline to integrate with NVD CVE API
+
+- **Author(s):** Yi Li <yli3@redhat.com>
+- **Created:** [2023-10-10]
+
+## Status
+
+Accepted
+
+## Context
+
+The National Vulnerability Database (NVD) has announced the retirement of its JSON feeds by December 15, 2023. This creates the need to change our existing CVSS data updater pipeline for Scanner V4, which currently relies on these JSON feeds.
+
+The NVD JSON feed offers bundled data on a yearly basis, spanning from 2002 to the present year. A benefit is that only two requests are needed in the workflow for each year (one for the meta file and one for the data bundle). However, a limitation is the absence of filtering capabilities, such as by CVSS score or modification date. In contrast, the NVD API allows data retrieval through specific HTTP URL parameters, such as publish date, modified date, or CVSS score, enabling more precise data acquisition.
+
+## Decision
+
+The updated NVD CVSS GitHub Workflow (CI) has transitioned to sourcing data from the NVD CVE API, rather than the NVD JSON feeds, ensuring the timely update of the NVD CVSS data bundle in Google Storage. The workflow is designed to download data by specifying a range of start and end dates, spanning from the year 2002 to the current year. For instance, a command listed below could be used:
+```
+curl "https://services.nvd.nist.gov/rest/json/cves/2.0/?pubStartDate=2021-08-04T00:00:00.000&pubEndDate=2021-10-22T00:00:00.000&startIndex=0" > 2021-test.json
+``` 
+In every 120-day range, it iteratively exhausts the startIndex to capture all associated data in that date range. After all json files downloaded, they will be compressed to one data bundle as one single compressed file. We choose start and end date to be 120 days of range because this is the maximum range that NVD supports.
```

Comment:
Instead of 120 days, I think it'd be easier to break the year down into 4 parts because it's easier to reason about it:

* January 1 - April 1
* April 1 - July 1
* July 1 - October 1
* October 1 - January 1

From here, it's still possible there are more than 2000 vulns in this range, so I'm picturing file that look like this:

`2023-A-B` where A = range number and B = index number

For example:

`2023-2-3` = vulns from 2023 from the July - October range and this is the 4th call within that range.

Does that make sense?

Reply:
I agree with this part. 
> January 1 - April 1
> April 1 - July 1
> July 1 - October 1
> October 1 - January 1
But maybe we can name each file as <start date> - <start index>.json
such as 2020-04-01-0.json
Then use jq to consolidate all "vunerabilities" fields to one big json file for that year to be "combine_2020.json"

Like:
```
if [ "$file_count" -gt 0 ]; then
        if jq -cs '{ vulnerabilities: map(.vulnerabilities) | add }' "$dir_name"/*"${year}"*.json > "$dir_name/combined_${year}.json"; then
            echo "The consolidated file size for $year is $(stat -c %s "$dir_name/combined_${year}.json") bytes."
```
Then remove the small sections of json data. 
So we will have 21 files in the tar file (2002 - 2023)